k8s/interpreter/100-interpreter-spec.yaml

```diff
@@ -78,6 +78,21 @@ spec:
   - name: spark-home
     emptyDir: {}
   {% endif %}
+  {% if zeppelin.k8s.interpreter.group.name == "flink" %}
+    volumeMounts:
+    - name: flink-home
+      mountPath: /flink
+  initContainers:
+  - name: flink-home-init
+    image: {{zeppelin.k8s.flink.container.image}}
+    command: ["sh", "-c", "cp -r /opt/flink/* /flink/"]
```

Comment:
I think you should change your `ZEPPELIN_K8S_CONTAINER_IMAGE` and add flink to /opt/flink.

Reply:
Actually, Flink is alreadly in `/opt/flink` directory in `ZEPPELIN_K8S_CONTAINER_IMAGE `. command here aims to mv flink dist file from `ZEPPELIN_K8S_CONTAINER_IMAGE` to volume `flink-home` which is mounted to '/flink' directory in initContainers. 

In interpreter container, I also mounted `flink-home` to '/flink' directory, maybe '/opt/flink' directory would be better. So I changed flink home to ‘/opt/flink’ in the new commit. [100-interpreter-spec.yaml Lin84](https://github.com/apache/zeppelin/pull/3884/files#diff-1b67202451861591b920d5d4a9e31ce8R84) ,  [zeppelin-server.yaml Line41](https://github.com/apache/zeppelin/pull/3884/files#diff-56ccb2e2c2617b27dbaae866d9431e51R41)

Reply:
> Actually, Flink is alreadly in /opt/flink directory in ZEPPELIN_K8S_CONTAINER_IMAGE

If Flink is already in your image (`ZEPPELIN_K8S_CONTAINER_IMAGE'), why do you need to copy Flink files that are in the initContainer into the interpreter container?

Reply:
sorry for wrong description. Actually , I mean flink is in '/opt/flink'  directory in image `ZEPPELIN_K8S_FLINK_CONTAINER_IMAGE`.

Reply:
Okay, thanks for clarifying. Why don't you extend your `ZEPPELIN_K8S_CONTAINER_IMAGE` with Flink under `/opt/flink`? If you extend your image, copying should not be necessary.

Reply:
that would need some manual work, including extend `ZEPPELIN_K8S_CONTAINER_IMAGE` and set flink home in interpreter, I want to do these work automatically.

Reply:
We can add Flink in PR #3859, just like we add Python and R packages.

Reply:
All in all you are right.
> add different kinds of interpreter related files together may make interpreter image too large.

Therefore we want to give the users the [responsibility](https://github.com/apache/zeppelin/pull/3859/files#diff-f55297559468e8c543ae5ab6ee356ec4R66) to build the best image for their use case. 
I think the copy process reduces the start time of the container and it makes the `interpreter-spec` more complicated. Today we have a special flink routine, tomorrow another one.

> it’s not flexible to change the interpreter related files，for example, if we need to change the flink version, we have to rebuild interpreter image

I think this is the biggest advantage and disadvantage of images. If you need to update any software version, including Flink, OpenJDK and operating system libraries, you have to rebuild your image. The advantage, you can make a nice rollback if your image contains all the application software. The disadvantage, of course, it that you have to rebuild your image, which takes quite a long time.

If you want to start certain interpreter images in your environment to reduce the image size, I recommend to create two or more interpreter images and inset a small if-then-else change [here](https://github.com/apache/zeppelin/blob/8ab7e9c473c5ce5b487bbab2a7b682f3fb165cab/k8s/interpreter/100-interpreter-spec.yaml#L45).

Reply:
Maybe it's not a good idea to copy files from images, I will close this PR, Thanks for explanation.